templates/common/upgrade/upgrade.go

```diff
@@ -96,28 +111,86 @@ type Params struct {
 	AllowDirtyTestOnly bool
 }
 
+type Result struct {
+	// If no upgrade was done because this installation of the template is
+	// already on the latest version, then this will be true and all other
+	// fields in this struct will have zero values.
+	AlreadyUpToDate bool
+
+	// Conflicts is the set of files that require manual intervention by the
+	// user to resolve a merge conflict. For example, there may be a file that
+	// was edited by the user and that edit conflicts with changes to that same
+	// file by the upgraded version of the template.
+	Conflicts []ActionTaken
+
+	// NonConflicts is the set of template output files that do NOT require any
+	// action by the user. Callers are free to ignore this.
+	//
+	// This is mutually exclusive with "Conflicts". Each file is in only one of
+	// the two lists.
+	NonConflicts []ActionTaken
+}
+
+// ActionTaken represents an output of the merge operation. Every file that's
+// part of the template output will have an ActionTaken that explains what the
+// merge algorithm decided to do for this file (e.g. Noop, EditEditConflict,
+// or other), and why.
+//
+// If there was a merge conflict, then files may have been renamed. See OursPath
+// and IncomingTemplatePath.
+type ActionTaken struct {
+	Action Action
+
+	// Explanation is a human-readable reason why the given Action was chosen
+	// for this path.
+	Explanation string
+
+	// This is the Path to the single file that this ActionTaken is about. It is
+	// always set.
+	//
+	// This is a relative path, starting from the directory where the template
+	// is installed.
+	Path string
+
+	// OursPath is only set for certain types of merge conflict. This is the
+	// path that the local file was renamed to that needs manual merge
+	// resolution.
+	//
+	// This is a relative path, starting from the directory where the template
+	// is installed.
+	OursPath string
+
+	// IncomingTemplatePath is only set for certain types of merge conflict.
+	// This is the path to the incoming template file that needs merge
+	// resolution.
+	//
+	// This is a relative path, starting from the directory where the template
+	// is installed.
+	IncomingTemplatePath string
+}
+
 // Upgrade takes a directory containing previously rendered template output and
 // updates it using the newest version of the template, which is pointed to by
 // the manifest file.
 //
 // Returns true if the upgrade occurred, or false if the upgrade was skipped
 // because we're already on the latest version of the template.
-func Upgrade(ctx context.Context, p *Params) (_ bool, rErr error) {
+func Upgrade(ctx context.Context, p *Params) (_ Result, rErr error) {
```

Comment:
Would it be easier if we return a pointer here?

Reply:
Yeah, done.